packages/flutter/lib/src/material/time_picker.dart

```diff
@@ -136,77 +485,147 @@ enum _TimePickerHeaderId {
+      double pieceHeight = 0.0;
+      for (final _TimePickerHeaderFragment fragment in piece.fragments) {
+        final Size childSize = childSizes[fragment.layoutId] = layoutChild(fragment.layoutId, constraints);
+        pieceHeight = math.max(pieceHeight, childSize.height);
+      }
+      pieceHeights.add(pieceHeight);
+      height += pieceHeight + margin;
+      // Delay application of margin until next piece because margin of the
+      // bottom-most piece should not contribute to the size.
+      margin = piece.bottomMargin;
+    }
 
-      // 11:57
-      //  --
-      // period
-      //
-      // The colon is centered horizontally, the entire layout is centered vertically.
-      // The "--" is a _kPeriodGap vertical gap.
-      case Orientation.landscape:
-        final double width = colonSize.width / 2.0 + minuteSize.width;
-        final double offset = math.max(0.0, size.width / 2.0 - width);
-        final double timeHeight = math.max(hourSize.height, colonSize.height);
-        final double height = timeHeight + _kPeriodGap + periodSize.height;
-        final double timeCenter = (size.height - height) / 2.0 + timeHeight / 2.0;
+    final _TimePickerHeaderPiece centrepiece = format.pieces[format.centrepieceIndex];
+    double y = (size.height - height) / 2.0;
+    for (int pieceIndex = 0; pieceIndex < format.pieces.length; pieceIndex += 1) {
+      if (pieceIndex != format.centrepieceIndex)
+        _positionPiece(size.width, y, childSizes, format.pieces[pieceIndex].fragments);
+      else
+        _positionPivoted(size.width, y, childSizes, centrepiece.fragments, centrepiece.pivotIndex);
 
-        double x = size.width - offset - minuteSize.width;
-        positionChild(_TimePickerHeaderId.minute, new Offset(x, timeCenter - minuteSize.height / 2.0));
+      y += pieceHeights[pieceIndex] + format.pieces[pieceIndex].bottomMargin;
+    }
+  }
 
-        x -= colonSize.width;
-        positionChild(_TimePickerHeaderId.colon, new Offset(x, timeCenter - colonSize.height / 2.0));
+  void _positionPivoted(double width, double y, Map<_TimePickerHeaderId, Size> childSizes, List<_TimePickerHeaderFragment> fragments, int pivotIndex) {
+    double tailWidth = childSizes[fragments[pivotIndex].layoutId].width / 2.0;
+    for (_TimePickerHeaderFragment fragment in fragments.skip(pivotIndex + 1)) {
+      tailWidth += childSizes[fragment.layoutId].width + fragment.leftMargin;
+    }
 
-        x -= hourSize.width;
-        positionChild(_TimePickerHeaderId.hour, new Offset(x, timeCenter - hourSize.height / 2.0));
+    double x = width / 2.0 + tailWidth;
+    x = math.min(x, width);
+    for (int i = fragments.length - 1; i >= 0; i -= 1) {
+      final _TimePickerHeaderFragment fragment = fragments[i];
+      final Size childSize = childSizes[fragment.layoutId];
+      x -= childSize.width;
+      positionChild(fragment.layoutId, new Offset(x, y - childSize.height / 2.0));
+      x -= fragment.leftMargin;
```

Comment:
We probably need RTL logic here

Reply:
Done, but I implemented it at the `_TimePickerHeaderPiece` level. If I understand correctly, RTL should not affect the positioning logic - everything is laid out pivoted around a specific fragment (the ":", "." or "h"). What it affects is the order in which `_TimePickerHeaderPiece`s appear.

Example:

```
"h:mm a":
  LTR: [piece("h:mm"), piece("a")]
  RTL: [piece("a"), piece("h:mm")]
```

In both cases everything is laid out pivoted around fragment ":":

```
LTR:
------|------
|    h:mm a |
------|------
      ^
    center

RTL:
------|------
|  a h:mm   |
------|------
      ^
    center

And in portrait mode:

LTR:
----|----
|  h:mm |
|   a   |
----|----
    ^
  center

RTL:
----|----
|   a   |
|  h:mm |
----|----
    ^
  center
```